RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc

```diff
@@ -0,0 +1,342 @@
+        if (!withHalo && (! spot.isCore() ))
+            continue;
+        if (initialCalculation) {
+            // initial calculation, take only core hits
+            if ( ! spot.isCore() ) continue;
+            //std::cout << " Multiplicity " << spot.multiplicity() << " " << spot.row()[0] << " " ;
+            //std::cout << spot.row()[1] << " " << spot.row()[2] << std::endl;
+            layers.insert(spot.layer());
+            for (int i = 0; i < spot.multiplicity(); ++i)
+                pca_->AddRow(spot.row());
+        }
+        else { // use a cylinder, include all hits
+            math::XYZPoint local = trans_(Point( spot.row()[0],spot.row()[1],spot.row()[2]));
+            if (local.Perp2() > radius2) continue;
+            layers.insert(spot.layer());
+            for (int i = 0; i < spot.multiplicity(); ++i)
+                pca_->AddRow(spot.row());
+        }
+    }
+    if (debug_)
+        std::cout << " Nlayers " << layers.size() << std::endl;
+    if (layers.size() < 3) {
+        pcaIteration_ = -1;
+        return;
+    }
+    pca_->MakePrincipals();
+    ++pcaIteration_;
+    const TVectorD means = *(pca_->GetMeanValues());
+    const TMatrixD eigens = *(pca_->GetEigenVectors());
+    const TVectorD eigenVals= *(pca_->GetEigenValues());
+    const TVectorD sigmas = *(pca_->GetSigmas());
+
+    barycenter_ = math::XYZPoint(means[0], means[1], means[2]);
+    axis_ = math::XYZVector(eigens(0, 0), eigens(1, 0), eigens(2, 0));
+    if (axis_.z() * barycenter_.z() < 0.0) {
+        axis_ = math::XYZVector(-eigens(0, 0), -eigens(1, 0), -eigens(2, 0));
+    }
+}
+
+ void EGammaPCAHelper::computeShowerWidth(float radius, bool withHalo){
+    sigu_ = 0.;
+    sigv_ = 0.;
+    sigp_ = 0.;
+    sige_ = 0.;
+    double cyl_ene = 0.;
+
+    float radius2 = radius * radius;
+    for ( auto spot : theSpots_) {
+        Point globalPoint(spot.row()[0],spot.row()[1],spot.row()[2]);
+        math::XYZPoint local = trans_(globalPoint);
+        if (local.Perp2() > radius2) continue;
+
+        // Select halo hits or not
+        if (withHalo && spot.fraction() < 0) continue;
+        if (!withHalo && !(spot.isCore())) continue;
+
+        sige_ += (globalPoint.eta() - theCluster_->eta()) * (globalPoint.eta() - theCluster_->eta()) * spot.energy();
+        sigp_ += deltaPhi(globalPoint.phi(), theCluster_->phi()) * deltaPhi(globalPoint.phi(), theCluster_->phi()) *
+              spot.energy();
+
+        sigu_ += local.x() * local.x() * spot.energy();
+        sigv_ += local.y() * local.y() * spot.energy();
+        cyl_ene += spot.energy();
+    }
+
+  if (cyl_ene > 0.) {
+    sigu_ = sigu_ / cyl_ene;
+    sigv_ = sigv_ / cyl_ene;
+    sigp_ = sigp_ / cyl_ene;
+    sige_ = sige_ / cyl_ene;
+  }
+  sigu_ = std::sqrt(sigu_);
+  sigv_ = std::sqrt(sigv_);
+  sigp_ = std::sqrt(sigp_);
+  sige_ = std::sqrt(sige_);
+}
+
+bool EGammaPCAHelper::checkIteration() const {
+    if (pcaIteration_ == 0) {
+        if(debug_) {
```

Comment:
I don't think the `return false` should be inside the scope of `if(debug_)` here (it isn't in the other cases below).